include/material/norsand.tcc

```diff
@@ -0,0 +1,627 @@
+
+  // Estimate maximum image pressure
+  double p_image_max = (mean_p + p_cohesion) *
+                       std::pow((1 + D_min * N_ / M_theta), ((N_ - 1) / N_));
+
+  // Compute derivatives
+  // Compute dF / dp
+  double dF_dp = -1. * M_theta / N_ *
+                 (1 - std::pow(((mean_p + p_cohesion) / (p_image + p_cohesion)),
+                               (N_ / (1 - N_))));
+
+  // Compute dp / dsigma
+  Vector6d dp_dsigma = Vector6d::Zero();
+  dp_dsigma(0) = 1. / 3.;
+  dp_dsigma(1) = 1. / 3.;
+  dp_dsigma(2) = 1. / 3.;
+
+  // Compute dF / dq
+  const double dF_dq = 1.;
+
+  // Compute the deviatoric stress
+  Vector6d dev_stress = Vector6d::Zero();
+  dev_stress(0) = stress(0) - mean_p;
+  dev_stress(1) = stress(1) - mean_p;
+  dev_stress(2) = stress(2) - mean_p;
+  dev_stress(3) = stress(3);
+  if (Tdim == 3) {
+    dev_stress(4) = stress(4);
+    dev_stress(5) = stress(5);
+  }
+
+  // Compute dq / dsigma
+  Vector6d dq_dsigma = Vector6d::Zero();
+  dq_dsigma(0) = 3. / 2. / deviatoric_q * dev_stress(0);
+  dq_dsigma(1) = 3. / 2. / deviatoric_q * dev_stress(1);
+  dq_dsigma(2) = 3. / 2. / deviatoric_q * dev_stress(2);
+  dq_dsigma(3) = 3. / deviatoric_q * dev_stress(3);
+  if (Tdim == 3) {
+    dq_dsigma(4) = 3. / deviatoric_q * dev_stress(4);
+    dq_dsigma(5) = 3. / deviatoric_q * dev_stress(5);
+  }
+
+  const double lode_angle = (*state_vars)["lode_angle"];
+  const double j2 = (*state_vars)["j2"];
+  const double j3 = (*state_vars)["j3"];
+  const double sin_lode_angle = sin(3. / 2. * lode_angle + M_PI / 4.);
+
+  // Compute dF / dM
+  double dF_dM = -1.0 / N_ *
+                 (1 + (N_ - 1) * std::pow(((mean_p + p_cohesion) /
+                                           (p_image + p_cohesion + p_dilation)),
+                                          (N_ / (1 - N_))));
+
+  // Compute dM / dtehta
+  const double dM_dtheta =
+      3. / 2. * std::pow(Mtc_, 2) / (3. + Mtc_) * sin_lode_angle;
+
+  // Compute dj2 / dsigma
+  Vector6d dj2_dsigma = dev_stress;
+
+  // Compute dj3 / dsigma
+  Eigen::Matrix<double, 3, 1> dev1;
+  dev1(0) = dev_stress(0);
+  dev1(1) = dev_stress(3);
+  dev1(2) = dev_stress(5);
+  Eigen::Matrix<double, 3, 1> dev2;
+  dev2(0) = dev_stress(3);
+  dev2(1) = dev_stress(1);
+  dev2(2) = dev_stress(4);
+  Eigen::Matrix<double, 3, 1> dev3;
+  dev3(0) = dev_stress(5);
+  dev3(1) = dev_stress(4);
+  dev3(2) = dev_stress(2);
+  Vector6d dj3_dsigma = Vector6d::Zero();
+  dj3_dsigma(0) = dev1.dot(dev1) - (2. / 3.) * j2;
+  dj3_dsigma(1) = dev2.dot(dev2) - (2. / 3.) * j2;
+  dj3_dsigma(2) = dev3.dot(dev3) - (2. / 3.) * j2;
+  dj3_dsigma(3) = dev1.dot(dev2);
+  if (Tdim == 3) {
+    dj3_dsigma(4) = dev2.dot(dev3);
+    dj3_dsigma(5) = dev1.dot(dev3);
+  }
+
+  // Compute dtheta / dsigma
+  Vector6d dtheta_dsigma = std::sqrt(3.) / 2. / cos(3 * lode_angle) /
+                           std::pow(j2, 1.5) *
+                           (dj3_dsigma - 3. / 2. * j3 / j2 * dj2_dsigma);
+  if (Tdim == 2) {
```

Comment:
Either handle 2D as special cases or 3D, there seems to be both?

Reply:
I think I would handle 2D as special cases. I have changed the previous lines